common/common.go

```diff
@@ -679,3 +679,39 @@ func (c *Counter) IncrementAndGet() int64 {
 	}
 	return newID
 }
+
+// ElementProcessor defines the function signature for processing an individual element with its index.
+// Index required in old subscription matching // TODO: Remove this when gateio templates are updated to all assets.
+type ElementProcessor[E any] func(index int, element E) error
+
+// ThrottledBatch takes a slice of elements and processes them in batches of `batchSize` concurrently.
+// For example, if batchSize = 10 and list has 100 elements, 10 goroutines will process 10 elements concurrently
+// in each batch. Each batch completes before the next batch begins.
+// `process` is a function called for each individual element with its index and value.
+func ThrottledBatch[S ~[]E, E any](batchSize int, list S, process ElementProcessor[E]) error {
```

Comment:
I feel like these names is something you've already discussed with gk. I don't think `ThrottledBatch` works here. You may wish to have a larger name for both elements to make it clearer.

My understanding of what you're saying is that this is `SequentialBatchConcurrentProcessing` Whereas the other function is `ConcurrentBatchSequentialProcessing`
But even with all the commentary, I'm not sure if I have the right of it.

Reply:
But even then, much like `ThrottleBatch` doesn't throttle, `BatchProcessor` doesn't actually mean it does anything sequentially. Its passed to a function via goroutines which may also do it concurrently.

Since I'm having such issue understanding what you're doing, I'm going to ask you to provide different names because I can't help

Reply:
Don't look at me for naming 😆. The only issue it rectifies is for an orderbook you need to suppress (throttle) the outbound requests because the websocket frame gets clipped in an orderbook snapshot context. So processing one block of 100 subscriptions and wait until that is done before going on to the next. Just note this is not a 429 issue as outbound subscriptions can occur without any rate limiting its only on the return leg and the amount of data is smashing against gorillas websocket frame. `SequentialBatchConcurrentProcessing` and `ConcurrentBatchSequentialProcessing` works for me. But I will get @gbjk to comment as you guys can whipsaw back and forth on this.